src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
 
         public abstract TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument);
 
-        public static IOperation CreateOperationNone(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        protected void SetParentOperation(IOperation parent)
         {
-            return new NoneOperation(node, constantValue, getChildren);
+            var result = Interlocked.CompareExchange(ref _parentDoNotAccessDirectly, parent, null);
+
+            // tree must belong to same semantic model if parent is given
+            Debug.Assert(parent == null || ((Operation)parent).SemanticModel == SemanticModel);
+
+            // make sure given parent and one we already have is same if we have one already
+            Debug.Assert(result == null || result == parent);
         }
 
-        private class NoneOperation : IOperation
+        /// <summary>
+        /// Create <see cref="IOperation"/> of <see cref="OperationKind.None"/> with explicit children
+        /// 
+        /// Use this to create IOperation when we don't have proper specific IOperation yet for given language construct
+        /// </summary>
+        public static IOperation CreateOperationNone(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
         {
-            private readonly Func<ImmutableArray<IOperation>> _getChildren;
+            return new NoneOperation(semanticModel, node, constantValue, getChildren);
+        }
+
+        public static T SetParentOperation<T>(T operation, IOperation parent) where T : IOperation
+        {
+            // explicit cast is not allowed, so using "as" instead
```

Comment:
I get this
Error	CS0030	Cannot convert type 'T' to 'Microsoft.CodeAnalysis.Operation'	

I can do this (Operation)(object)operation if you prefer that.